### PR TITLE
improvement(nemesis.py): calculate latency for perf

### PR DESCRIPTION
**Pull request by fgelcer**
Added a function that will calculate all the performance
latency we want to show.
Created a decorator, and added it to the parts of the
nemesis we want to measure the latencies.
Few of those parts were added to local functions
created inside their own nemesis functions.

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by fgelcer** [comment]
must be merged after #2426 will be merged.

**Comment by bentsi** [comment]
is this Draft?

**Comment by roydahan** [comment]
1. Fix conflicts.
2. Organize your commits.
3. Send a followup PR for storing and reporting (different task).

File: test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml
Comment (roydahan) [inline test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml:7]: Try to change it to sufficient time of prepare + 2 cycles of nemesis.

**Comment by roydahan** [comment]
Add also the pipeline job.

**Comment by fgelcer** [comment]
> Add also the pipeline job.

there is already a pipeline:
`jenkins-pipelines/perf-regression-latency-250gb-with-nemesis.jenkinsfile`

**Comment by fgelcer** [comment]
@roydahan 
this is a job that passed (last one)... job after this one failed with ENOSP of Jenkins, and the next one, with another issue that is being cleared out.

**Comment by roydahan** [comment]
@bentsi / @fruch can one of you review it?
We need it pretty urgent.

**Comment by fgelcer** [comment]
after chatting with @roydahan , fixed the Prometheus DB step, as for my test, it returned too many entries, and Prometheus DB blocks responses with more than 5M entries (so when i send a step bigger than the default, it returns small enough entries to get Prometheus to work).
other than that, emails are being sent, but there is still something (in the `TearDown()`) that is failing the Jenkins job.

**Comment by fgelcer** [comment]
> after chatting with @roydahan , fixed the Prometheus DB step, as for my test, it returned too many entries, and Prometheus DB blocks responses with more than 5M entries (so when i send a step bigger than the default, it returns small enough entries to get Prometheus to work).
> other than that, emails are being sent, but there is still something (in the `TearDown()`) that is failing the Jenkins job.

@fruch , last test passed!! you can see it [here](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/latency_automatic_calculation_mixed/)

now the only thing to do is, after this PR is approved, and prior to merging it, is to change the `docker/env/version` to its final format.

**Comment by fruch** [comment]
> > after chatting with @roydahan , fixed the Prometheus DB step, as for my test, it returned too many entries, and Prometheus DB blocks responses with more than 5M entries (so when i send a step bigger than the default, it returns small enough entries to get Prometheus to work).
> > other than that, emails are being sent, but there is still something (in the `TearDown()`) that is failing the Jenkins job.
> 
> @fruch , last test passed!! you can see it [here](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/latency_automatic_calculation_mixed/)
> 
> now the only thing to do is, after this PR is approved, and prior to merging it, is to change the `docker/env/version` to its final format.

@fgelcer I wasn't following this PR very closely, but I don't know how docker is related here, please explain

**Comment by fgelcer** [comment]
> > > after chatting with @roydahan , fixed the Prometheus DB step, as for my test, it returned too many entries, and Prometheus DB blocks responses with more than 5M entries (so when i send a step bigger than the default, it returns small enough entries to get Prometheus to work).
> > > other than that, emails are being sent, but there is still something (in the `TearDown()`) that is failing the Jenkins job.
> > 
> > 
> > @fruch , last test passed!! you can see it [here](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/latency_automatic_calculation_mixed/)
> > now the only thing to do is, after this PR is approved, and prior to merging it, is to change the `docker/env/version` to its final format.
> 
> @fgelcer I wasn't following this PR very closely, but I don't know how docker is related here, please explain

ah sorry, mixed it with the other PR i'm currently working on...
so this PR is ready to merge and no other changes are requested (so far).

**Comment by fgelcer** [comment]
@roydahan @fruch ping

File: sdcm/nemesis.py
Comment (roydahan) [inline sdcm/nemesis.py:586]: @fgelcer how does this decorator affects regular longevities?

**Comment by roydahan** [comment]
Before approving, the following regression tests with this branch need to be tested:
1. longevity short & long (one of each).
2. Performance test - 1 latency & 1 throughput.

File: sdcm/nemesis.py
Comment (fgelcer) [reply sdcm/nemesis.py:586]: @roydahan , this decorator will collect the latency metrics for this function, but nothing will be done with it in the end (I will check if it will be saved together with other data to ES)

**Comment by fgelcer** [comment]
> Before approving, the following regression tests with this branch need to be tested:
> 
> 1. longevity short & long (one of each).
> 2. Performance test - 1 latency & 1 throughput.

do you have any preference for a long longevity?

**Comment by roydahan** [comment]
Start with short one of 3h then if it passes we can discuss other.

**Comment by roydahan** [comment]
@fruch I need you to review the part of how this PR stores latency of nemesis in ES.

File: sdcm/utils/decorators.py
Comment (fruch) [inline sdcm/utils/decorators.py:156]: Why the import is it inside the function ?

File: sdcm/utils/latency.py
Comment (fruch) [inline sdcm/utils/latency.py:28]: why there a sleep here ? What are we waiting for ?

File: sdcm/utils/decorators.py
Comment (fgelcer) [reply sdcm/utils/decorators.py:156]: circular import

File: sdcm/utils/latency.py
Comment (fgelcer) [reply sdcm/utils/latency.py:28]: to give the time to the Prometheus connection to be open and functional

Comment (fruch) [reply sdcm/utils/latency.py:28]: Should be part of PrometheusDBStats object, and there isn't any persistent connection that I'm aware of it in there

File: sdcm/utils/decorators.py
Comment (fruch) [reply sdcm/utils/decorators.py:156]: Document it, otherwise some might try moving it

**Comment by fruch** [comment]
> @fruch I need you to review the part of how this PR stores latency of nemesis in ES.

@fgelcer please supply an example of an item saved to ES, from the code is very very hard to see the structure being saved

File: sdcm/utils/latency.py
Comment (fgelcer) [reply sdcm/utils/latency.py:28]: so you mean here to add this delay while creating the `PrometheusDBStats` object?

**Comment by fgelcer** [comment]
> > @fruch I need you to review the part of how this PR stores latency of nemesis in ES.
> 
> @fgelcer please supply an example of an item saved to ES, from the code is very very hard to see the structure being saved

this ES document ID has the results:
`dc105daf-7213-483b-bb67-017aa8699bd4_20201117_105045_308295`

they are nested under `latency_during_ops` json key object

**Comment by fgelcer** [comment]
> > > @fruch I need you to review the part of how this PR stores latency of nemesis in ES.
> > 
> > 
> > @fgelcer please supply an example of an item saved to ES, from the code is very very hard to see the structure being saved
> 
> this ES document ID has the results:
> `dc105daf-7213-483b-bb67-017aa8699bd4_20201117_105045_308295`
> 
> they are nested under `latency_during_ops` json key object

after we discussed the matter of the limit of keys ES has, i'm testing now a solution to the key added for each node, to be the node index, instead of IP address.
hope to have positive results quite soon

File: sdcm/utils/latency.py
Comment (aleksbykov) [inline sdcm/utils/latency.py:33]: WHy not monitor_node.ip_address?

File: sdcm/utils/latency.py
Comment (aleksbykov) [inline sdcm/utils/latency.py:43]: You don't need here the `step` parameter, do you?

File: sdcm/utils/latency.py
Comment (aleksbykov) [inline sdcm/utils/latency.py:46]: not clear names `lst` and `response`. May be latency_values_lst and lat_value?

File: test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml
Comment (roydahan) [inline test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml:7]: Why 350m?
It's almost 6 hours, how long the nemesis cycle takes?

File: sdcm/utils/latency.py
Comment (fgelcer) [reply sdcm/utils/latency.py:37]: not need in here, as this is of a short period (start and and here refer to start and end of a nemesis)

File: performance_regression_test.py
Comment (fgelcer) [reply performance_regression_test.py:239]: changed to `scrap_metrics_step`

File: sdcm/utils/latency.py
Comment (fgelcer) [reply sdcm/utils/latency.py:27]: updated

File: sdcm/utils/latency.py
Comment (fgelcer) [reply sdcm/utils/latency.py]: updated

File: test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml
Comment (fgelcer) [reply test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml:7]: a full cycle (in the latest master run) took ~5 hours, but this stress is not limited to it... it also includes a period (30 minutes in this test) before the nemesis is called for the 1st time (and the nemesis includes the steady state and the whole sequence)

**Review by fruch — APPROVED** [review]
node by name, seems to be o.k. now from my POV

**Comment by fgelcer** [comment]
@roydahan , conflicts fixed for the 2nd time in 10 min... if already approved, please merge

**Comment by roydahan** [comment]
@fgelcer still conflicts

File: test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml
Comment (roydahan) [inline test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml:36]: change back to scylla-perf-results instead of QA

File: test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml
Comment (roydahan) [inline test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml:12]: please change  the instance_type_loader to c5  instead of c4.

**Review by roydahan — CHANGES_REQUESTED** [review]
Small changes in test yaml.

**Comment by fgelcer** [comment]
> @fgelcer still conflicts

hope that, until next year when you will click in the merge button there will not be more conflicts

**Review by roydahan — APPROVED** [review]

**Comment by roydahan** [comment]
Merging this because it becomes an endless loop of work.
I'm creating branch-perf-v10 and we will backport fixes there.

**Comment by bentsi** [comment]
this PR broke `_add_and_init_new_cluster_node`... that caused the failure of all the Nemesis that use it...

**Comment by fgelcer** [comment]
> this PR broke `_add_and_init_new_cluster_node`... that caused the failure of all the Nemesis that use it...

can you please provide a failed test?

**Pull request merged**
